### PR TITLE
Add pulp_label_filter for filtering by pulp_labels

### LINKED ISSUE #8067: upgrading to 3.117.1 fails without saying which Distribution base_paths are violating constraint
**Version**
3.117.1
**Describe the bug**

After failing to migrate to 3.117.0 with deadlocks during migrations the upgrade to 3.117.1 fails with 

```
+ set -e
+ /usr/local/lib/pulp/bin/pulpcore-manager migrate --noinput
Operations to perform:
  Apply all migrations: admin, auth, certguard, container, contenttypes, core, file, hugging_face, maven, npm, python, rpm, service, sessions
Running migrations:
  Applying core.0156_alter_contentartifact_relative_path_and_more...Traceback (most recent call last):
  File "/usr/local/lib/pulp/lib64/python3.11/site-packages/django/db/backends/utils.py", line 103, in _execute
    return self.cursor.execute(sql)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/pulp/lib64/python3.11/site-packages/psycopg/cursor.py", line 117, in execute
    raise ex.with_traceback(None)
psycopg.errors.CheckViolation: value for domain relative_path violates check constraint "relative_path_check"

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/pulp/bin/pulpcore-manager", line 6, in <module>
    sys.exit(manage())
             ^^^^^^^^
  File "/usr/local/lib/pulp/lib64/python3.11/site-packages/pulpcore/app/manage.py", line 11, in manage
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/pulp/lib64/python3.11/site-packages/django/core/management/__init__.py", line 442, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/pulp/lib64/python3.11/site-packages/django/core/management/__init__.py", line 436, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/pulp/lib64/python3.11/site-packages/django/core/management/base.py", line 420, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/pulp/lib64/python3.11/site-packages/django/core/management/base.py", line 464, in execute
    output = self.handle(*args, **options)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/pulp/lib64/python3.11/site-packages/django/core/management/base.py", line 111, in wrapper
    res = handle_func(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/pulp/lib64/python3.11/site-packages/django/core/management/commands/migrate.py", line 353, in handle
    post_migrate_state = executor.migrate(
                         ^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/pulp/lib64/python3.11/site-packages/django/db/migrations/executor.py", line 135, in migrate
    state = self._migrate_all_forwards(
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/pulp/lib64/python3.11/site-packages/django/db/migrations/executor.py", line 167, in _migrate_all_forwards
    state = self.apply_migration(
            ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/pulp/lib64/python3.11/site-packages/django/db/migrations/executor.py", line 255, in apply_migration
    state = migration.apply(state, schema_editor)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/pulp/lib64/python3.11/site-packages/django/db/migrations/migration.py", line 132, in apply
    operation.database_forwards(
  File "/usr/local/lib/pulp/lib64/python3.11/site-packages/django/db/migrations/operations/fields.py", line 241, in database_forwards
    schema_editor.alter_field(from_model, from_field, to_field)
  File "/usr/local/lib/pulp/lib64/python3.11/site-packages/django/db/backends/base/schema.py", line 912, in alter_field
    self._alter_field(
  File "/usr/local/lib/pulp/lib64/python3.11/site-packages/django/db/backends/postgresql/schema.py", line 274, in _alter_field
    super()._alter_field(
  File "/usr/local/lib/pulp/lib64/python3.11/site-packages/django/db/backends/base/schema.py", line 1165, in _alter_field
    self.execute(
  File "/usr/local/lib/pulp/lib64/python3.11/site-packages/django/db/backends/postgresql/schema.py", line 48, in execute
    return super().execute(sql, None)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/pulp/lib64/python3.11/site-packages/django/db/backends/base/schema.py", line 204, in execute
    cursor.execute(sql, params)
  File "/usr/local/lib/pulp/lib64/python3.11/site-packages/sentry_sdk/utils.py", line 1886, in runner
    return sentry_patched_function(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/pulp/lib64/python3.11/site-packages/sentry_sdk/integrations/django/__init__.py", line 832, in execute
    result = real_execute(self, sql, params)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/pulp/lib64/python3.11/site-packages/django/db/backends/utils.py", line 79, in execute
    return self._execute_with_wrappers(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/pulp/lib64/python3.11/site-packages/django/db/backends/utils.py", line 92, in _execute_with_wrappers
    return executor(sql, params, many, context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/pulp/lib64/python3.11/site-packages/django/db/backends/utils.py", line 100, in _execute
    with self.db.wrap_database_errors:
  File "/usr/local/lib/pulp/lib64/python3.11/site-packages/django/db/utils.py", line 91, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/usr/local/lib/pulp/lib64/python3.11/site-packages/django/db/backends/utils.py", line 103, in _execute
    return self.cursor.execute(sql)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/pulp/lib64/python3.11/site-packages/psycopg/cursor.py", line 117, in execute
    raise ex.with_traceback(None)
django.db.utils.IntegrityError: value for domain relative_path violates check constraint "relative_path_check"

```

### PR DESCRIPTION
**Pull request by daviddavis**
fixes #8067

Please be sure you have read our documentation on creating PRs:
https://docs.pulpproject.org/contributing/pull-request-walkthrough.html

### CODE REVIEW

File: pulpcore/tests/functional/api/using_plugin/test_labels.py [hunk 1]
```diff
@@ -116,3 +138,71 @@ def test_invalid_labels(self):
...
+        """Create an API client."""
+        self.client = FileApiClient(self.cfg.get_bindings_config())
+        self.repo_api = RepositoriesFileApi(self.client)
+
+    def _create_repo(self, labels={}):
+        attrs = {"name": str(uuid4())}
+        if labels:
+            attrs["pulp_labels"] = labels
+        self.repo = self.repo_api.create(attrs)
+        self.addCleanup(self.repo_api.delete, self.repo.pulp_href)
```
Comment (mdellweg) [inline pulpcore/tests/functional/api/using_plugin/test_labels.py:162]: This is already in the base class, right?

**Comment by mdellweg** [comment]
I like this! Is it possible to have some thing like "env in ['prod', 'staging']"?

**Comment by daviddavis** [comment]
> I like this! Is it possible to have some thing like "env in ['prod', 'staging']"?

The kubernetes label select offers this. Our stakeholders said we could get away with not having this initially. But that reminds me: I meant to file an issue but forgot about it. Here we go: https://pulp.plan.io/issues/8174

**Review by mdellweg — APPROVED** [review]

File: pulpcore/app/viewsets/custom_filters.py [hunk 2]
```diff
@@ -274,3 +276,56 @@ def filter(self, qs, value):
...
+    """Filter to get resources that match a label filter string."""
+
+    def __init__(self, *args, **kwargs):
+        kwargs.setdefault("help_text", _("Filter labels by search string"))
+        super().__init__(*args, **kwargs)
+
+    def filter(self, qs, value):
+        """
+        Args:
+            qs (django.db.models.query.QuerySet): The RepositoryVersion Queryset
```
Comment (goosemania) [inline pulpcore/app/viewsets/custom_filters.py:291]: Did you mean `The Model Queryset`?

File: pulpcore/app/viewsets/custom_filters.py [hunk 3]
```diff
@@ -274,3 +276,56 @@ def filter(self, qs, value):
...
+
+        Raises:
+            rest_framework.exceptions.ValidationError: on invalid search string
+        """
+        if value is None:
+            # user didn't supply a value
+            return qs
+
+        for term in value.split(","):
+            match = re.match(r"(!?[\w\s]+)(=|!=|~)?(.*)?", term)
```
Comment (goosemania) [inline pulpcore/app/viewsets/custom_filters.py:305]: :+1:

File: pulpcore/app/viewsets/custom_filters.py [hunk 4]
```diff
@@ -274,3 +276,56 @@ def filter(self, qs, value):
...
+            if not match:
+                raise DRFValidationError(_("Invalid search term: '{}'.").format(term))
+            key, op, val = match.groups()
```
Comment (goosemania) [inline pulpcore/app/viewsets/custom_filters.py:308]: It seems to me that one can get value=None when operation is specified.
E.g. `"environment="`.
Should cover it in tests?
And what would be the behaviour? Match everything where this label exists and set to None?
What about `environment~`?  Not sure how `icontains` react to the None value.

Comment (daviddavis) [reply pulpcore/app/viewsets/custom_filters.py:308]: `val` gets set to `''` when specifying `environment=` or `environment~`. So `environment=` would match label with environment with value of `''` while `environment~` matches any label with key `environment`. I can add some tests for these cases.

Comment (goosemania) [reply pulpcore/app/viewsets/custom_filters.py:308]: Thanks, +1 to add tests

**Review by goosemania — APPROVED** [review]
Thanks!

**Pull request merged**